### ISSUE TITLE
Per-user settings/workspace

### ISSUE THREAD

**Issue by davidbrochart** [jupyter-server/jupyverse#61]
### Problem

In Jupyverse, settings and workspace are stored alongside the user information, in a database.
When jupyverse is launched with `--authenticator.mode=noauth`, there is no authentication and the whole API is unprotected. Naturally, we want settings and workspace to be stored with this unique, "global user", so that users get their settings back the next time they launch jupyverse.
When jupyverse is launched with `--authenticator.mode=token`, a token is created and a user is registered in the DB with this token as its ID. When jupyverse is stopped, the user is removed from the DB. What this means is that this is a new user each time, and settings won't persist between jupyverse launches.

### Proposed Solution

Maybe each token-user should inherit the unauthenticated, "global user" settings and workspace, by copying them when the token-user is registered? That would at least guarantee some persistence for settings and workspace. Similarly, settings and workspace could be copied back to the "global user" when the token-user jupyverse is stopped?
If several unauthenticated-user or token-user jupyverse are launched at the same time, this is obviously a source of conflicts, but the main use-case of these two modes are really a single user launching a single server at a time.

**Comment by fcollonval** [comment]
> Naturally, we want settings and workspace to be stored with this unique, "global user", so that users get their settings back the next time they launch jupyverse.

I don't sure this should be the default behaviour for `noauth` case. This case should never be used except for testing or some very exotic configuration. But the single server case should always be using a token (this is the current behavior of Jupyter server for security reasons). So I would say that the settings and workspaces should never be persisted for the `noauth` case (although their should be enough persistence to have a coherent session).

Regarding the token case, this is probably more tricky. In particular using token in a Cloud served JupyterLab is probably different than using it for a local instance. So I'm wondering if the host name of the server should be used in such anonymous cases to distinguish local execution for which the user expect its settings and workspaces to be persistent when on a web service they should not be persisted.

**Comment by davidbrochart** [comment]
> So I would say that the settings and workspaces should never be persisted for the `noauth` case

Note that this is different from the current behavior of Jupyter Server, which persists settings and workspaces when authentication is disabled.

**Comment by fcollonval** [comment]
> Note that this is different from the current behavior of Jupyter Server, which persists settings and workspaces when authentication is disabled.

Yep and that brings lots of problems for integration tests. So changing that will help on that front.

**Comment by davidbrochart** [comment]
I am not opposed to change it, but if it's just a convenience for tests, we could also have a configuration to disable settings persistence.

**Comment by fcollonval** [comment]
I agree the primary concern should not be test convenience. But I am convinced that in this discussion the security aspects are missing. In particular sharing data between unknown users as default policy is usually a bad idea; for example the settings may contain secrets or personal information. So it will make sense to clarify the user story and security concerns behind each mode, something like:


| mode | story | security |
| --- | --- | --- |
| no auth | Testing | Nothing should be persisted once the session ends |
| token | Single server user | Data should not be accessible from remotes |
| | | Data should be persisted and served back based on the OS user (TBC) |
| auth | Deployed infrastructure | Data of the user are served if its identity is valid |
| | Invitation to a shared session | ? |
